### PR TITLE
Fix output property names in Python codegen.

### PR DESCRIPTION
**Pull request by pgavlin**
Some property names are mapped from their `camelCase` Pulumi name to a
`snake_case` Python name. This mapping is irregular, and only occurs for
resources properties and function calls.

Note that there's still more work to do here: this only fixes names on
the output side; the input side is still broken for nested resource
proprerties and function calls.

The underlying design--annotated types in `hcl2/model`--may need some
additional work in the future, but I _believe_ it's good enough for now.

### CODE REVIEW

File: pkg/codegen/hcl2/model/type.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ type Type interface {
 
 	AssignableFrom(src Type) bool
 	ConversionFrom(src Type) ConversionKind
+	GetAnnotations() []interface{}
```
Comment (pgavlin) [inline pkg/codegen/hcl2/model/type.go:40]: This should probably return something with an API more like that of `context.Context`: a logical map keyed by package-level variables.

File: pkg/codegen/hcl2/binder_schema.go [hunk 2]
```diff
@@ -124,22 +124,26 @@ func (b *binder) loadPackageSchema(name string) error {
 }
 
 // schemaTypeToType converts a schema.Type to a model Type.
-func schemaTypeToType(src schema.Type) model.Type {
+func (b *binder) schemaTypeToType(src schema.Type) (result model.Type) {
+	defer func() {
+		b.typeSchemas[result] = src
```
Comment (lblackstone) [inline pkg/codegen/hcl2/binder_schema.go:129]: This is an interesting approach. I guess the inline alternative would be some form of goto?

File: pkg/codegen/hcl2/model/type_list.go [hunk 3]
```diff
@@ -104,6 +104,10 @@ func (t *ListType) conversionFrom(src Type, unifying bool) ConversionKind {
 	})
 }
 
+func (t *ListType) GetAnnotations() []interface{} {
+	return nil
```
Comment (lblackstone) [inline pkg/codegen/hcl2/model/type_list.go:108]: Is this always going to be `nil`, or do we want to include a TODO, or placeholder message of some sort?

File: pkg/codegen/hcl2/program.go [hunk 4]
```diff
@@ -96,3 +98,18 @@ func (p *Program) NewDiagnosticWriter(w io.Writer, width uint, color bool) hcl.D
...
+	names := make([]string, 0, len(p.binder.packageSchemas))
+	for name := range p.binder.packageSchemas {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+
+	packages := make([]*schema.Package, len(names))
+	for i, name := range names {
+		packages[i] = p.binder.packageSchemas[name].schema
+	}
```
Comment (lblackstone) [inline pkg/codegen/hcl2/program.go:113]: Any reason not to match the format of the previous loop?
```suggestion
	packages := make([]*schema.Package, 0, len(names))
	for _, name := range names {
		packages = append(packages, p.binder.packageSchemas[name].schema)
	}
```

File: pkg/codegen/internal/test/testdata/aws-s3-logging.pp.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+
+const logs = new aws.s3.Bucket("logs", {});
+const bucket = new aws.s3.Bucket("bucket", {loggings: [{
+    targetBucket: logs.bucket,
+}]});
+export const targetBucket = bucket.loggings.apply(loggings => loggings[0].targetBucket);
```
Comment (lblackstone) [inline pkg/codegen/internal/test/testdata/aws-s3-logging.pp.ts:8]: Should this output be lifted?

File: pkg/codegen/schema/schema.go [hunk 6]
```diff
@@ -940,24 +953,26 @@ func bindDefaultValue(value interface{}, spec *DefaultSpec, typ Type) (*DefaultV
 	return dv, nil
 }
 
-func (t *types) bindProperties(properties map[string]PropertySpec, required []string) ([]*Property, error) {
+func (t *types) bindProperties(properties map[string]PropertySpec,
```
Comment (lblackstone) [inline pkg/codegen/schema/schema.go:956]: Could you add a docstring? Hard to follow what the output values are.

File: pkg/codegen/hcl2/program.go (hunk 4 above)
Comment (pgavlin) [reply pkg/codegen/hcl2/program.go:113]: Yes: I prefer to use explicit indices when they're trivially available. The previous loop can't use them because the expression being ranged over is a map.

File: pkg/codegen/internal/test/testdata/aws-s3-logging.pp.ts (hunk 5 above)
Comment (pgavlin) [reply pkg/codegen/internal/test/testdata/aws-s3-logging.pp.ts:8]: Yup, filed https://github.com/pulumi/pulumi/issues/4465

File: pkg/codegen/hcl2/binder_schema.go (hunk 2 above)
Comment (pgavlin) [reply pkg/codegen/hcl2/binder_schema.go:129]: Correct.

File: pkg/codegen/hcl2/model/type_list.go (hunk 3 above)
Comment (pgavlin) [reply pkg/codegen/hcl2/model/type_list.go:108]: This is going to be nil indefinitely. I'll file an issue for a proper design here; getting the type system to track annotations in a way that makes sense is tricky.

File: pkg/codegen/schema/schema.go (hunk 6 above)
Comment (pgavlin) [reply pkg/codegen/schema/schema.go:956]: Done.

File: pkg/codegen/hcl2/model/type_list.go (hunk 3 above)
Comment (pgavlin) [reply pkg/codegen/hcl2/model/type_list.go:108]: https://github.com/pulumi/pulumi/issues/4466
